### DISCUSSION TITLE
What are the 10 commands you need to be really productive?

### DISCUSSION THREAD

**Discussion by frosencrantz** [saulpw/visidata#1505] [General]
I just watched a video @saulpw did with @cpcloud https://youtu.be/RSDkLGjj3-E about using https://github.com/visidata/vdsql . Saul was on the keyboard. Thanks for doing the video, I'm looking forward to trying out vdsql.

Somewhere in the middle Saul says:
  > okay for my money there's actually only about 10 commands that you need  to be actually really productive

Saul what are the top 10-commands?  I'd also be interested in what other people think too.

**Comment by saulpw** [comment]
Everyone will probably have their own specific 10, but here are mine:

1. freq table: `Shift+F`
2. aggregation: `+` on columns before doing freq table
3. hide column: `-`
4. sort: `[` or `]`
5. select rows by example: `,`
6. bulk select/toggle/unselect rows: `gs`/`gt`/`gu`
7. filter selected rows only: `"` (`dup-sheet`)
8. create a new column by Python expr: `=`
9. bulk edit all selected: `ge`
10. commit updates to my personal database: `z Ctrl+S`

Honorable mention for `syscopy-cell` or `z Shift+Y`, that copies a cell to the system clipboard (so it can be pasted into another window).

**Reply by ajkerrigan** [reply]
> Everyone will probably have their own specific 10, but here are mine:

Yeah this feels pretty personal and at least somewhat task-dependent. I'd have some overlap with Saul's but with more of a skew toward slinging nested JSON. Maybe...

1. column-wise expansion of nested data (`expand-col*`, `contract-col*`, and built-in/custom variations on those)
1. row-wise expansion of nested data (`unfurl-col`)
1. create a new column by Python expr (`=`)
1. rename column (`^`)
1. hide column (`-`)
1. freq table (`Shift+F`)
1. sort `[` or `]`
1. select rows by example (`,`)
1. bulk select/toggle/unselect rows (`gs`/`gt`/`gu`)
1. filter selected rows only: (`"` / `dup-sheet`)

**Comment by frosencrantz** [comment]
Here’s my top ten/twenty list.  It is partly between Saul and  AJ.  Thanks @saulpw and AJ.


@ajkerrigan I’d be interested to hear what custom variations you have of expand-col/contract-col.  Is this your JMESpath plugin? 
https://github.com/ajkerrigan/visidata-plugins#vd_jmespath-evaluate-jmespath-expressions

This was the common top 7ish
- freq table: `Shift+F`
- hide column: `-`
- sort: `[` or `]`
- select rows by example: `,`
- bulk select/toggle/unselect rows: `gs`/`gt`/`gu`
- filter selected rows only: `"` (`dup-sheet`)
- create a new column by Python expr: `=`

Here's what I would add/agree with.
- Like AJ I also work with nested data, my top 3 cmds are `unfurl-col`/`pyobj-cell`/`expand-col`
- Help-sheets are very useful for me to find the longer tail commands I want to use Help Sheets `zCtrl-H` &  `gzCtrl+H`.   I find I'm using the menu system frequently.
- Understanding how to use the FreqSheet is very powerful (including the `ge` command, selection commands, and dive-row(s))
- And basic commands like `q` quit, Ctrl-S save are critical.


Wondering  if others have a top ten list of most powerful commands you use.

**Reply by ajkerrigan** [reply]
I definitely use the JMESpath stuff regularly, yeah. Not because I'm particularly in love with the language, it just pops up a lot in stuff I work on. As for other nested shenanigans...

* VisiData has a native `)`/`contract-col` command, I added a `g)`/`contract-cols` command to quickly contract all columns in the sheet.
* I have custom `z(` and `gz(` bindings to recursively expand one or all columns.
* When I was using vfake a lot, I added an `accordion-autofake` command that would expand all columns, run [autofake](https://github.com/ajkerrigan/visidata-plugins/#vfake_extensions-niche-addons-for-vfake), then contract them again. Not something I use often today, but it was very useful at the time :)!
* [Convenience commands](https://github.com/ajkerrigan/visidata-plugins/#kvpairs-toggle-values-between-lists-of-keyvalue-pairs-and-dicts) to toggle between dicts and arrays of key/value pairs.
* `zo` to open a cell with a specific loader. Mostly useful when a cell contains stringified json and I want to dig a bit, and don't necessarily need to do it for all rows in the column.

Good call not forgetting the basic commands! I don't usually edit cells directly, but that does seem like a key thing to know is possible :+1: . Your mention of basics got me thinking that `S` and `C` to pop open the Sheets and Columns sheets are pretty important too. And resizing/wrapping is super important if you're dealing with long values. Who would have thought... coming up with a top 10 is really hard :sweat_smile:

**Reply by frosencrantz** [reply]
@ajkerrigan  thanks for that great response!  I liked hearing about the custom commands you've created:

- I haven't had to use fake data, but that trick about expanding and faking the data seems like it could be very useful.
- `zo` is that what was suggested in https://github.com/saulpw/visidata/issues/465? 
- I’ve used your convenience functions for key/value pairs, and it was useful. I  was happy you shared that code, but I haven’t had to use it recently. 

Agreed on the basic commands. `S` is burned into finger memory. `C` is also close behind. Resize commands such as `_` and `z-` are very common in my usage.  Yes coming up with a top 10 is hard.  😃 

I think one thing I thought might be on @saulpw  original list was `W`/`M` pivot/melt commands.

Curious if @anjakefala  or anyone else has a top ten that they would like to share.

**Comment by anjakefala** [comment]
@frosencrantz 

> Curious if @anjakefala or anyone else has a top ten that they would like to share.

I have a serious answer to this question, that I will type up one day but right now my brain's answer is:

"My top command is `Zzzzzzzz` (longname `sleep-hours-n`)"
